torch/utils/data/graph.py

```diff
@@ -70,21 +72,22 @@ def reduce_hook(obj):
     return captured_connections
 
 
-def traverse(datapipe, only_datapipe=False):
-    cache: Set[DataPipe] = set()
+def traverse(datapipe: DataPipe, only_datapipe: bool = False) -> DataPipeGraph:
+    cache: Set[int] = set()
     return _traverse_helper(datapipe, only_datapipe, cache)
 
 
 # Add cache here to prevent infinite recursion on DataPipe
-def _traverse_helper(datapipe, only_datapipe, cache):
+def _traverse_helper(datapipe: DataPipe, only_datapipe: bool, cache: Set[int]) -> DataPipeGraph:
     if not isinstance(datapipe, (IterDataPipe, MapDataPipe)):
         raise RuntimeError("Expected `IterDataPipe` or `MapDataPipe`, but {} is found".format(type(datapipe)))
 
-    cache.add(datapipe)
+    dp_id = id(datapipe)
+    cache.add(id(dp_id))
```

Comment:
```suggestion
    cache.add(dp_id)
```
I think this is what you want and should fix the infinite recursion in CI.